src/reducers/shelf/filter.ts

```diff
@@ -0,0 +1,47 @@
+
+import {ADD_ONE_OF_FILTER, ADD_RANGE_FILTER, REMOVE_ONE_OF_FILTER, REMOVE_RANGE_FILTER} from '../../actions/filter';
+import {Action} from '../../actions/index';
+import {DEFAULT_SHELF_UNIT_SPEC, ShelfUnitSpec} from '../../models/shelf/spec';
+import {insertItemToArray, removeItemFromArray} from './spec';
+
+export function filterReducer(shelfSpec: Readonly<ShelfUnitSpec> = DEFAULT_SHELF_UNIT_SPEC,
+                              action: Action): ShelfUnitSpec {
+  switch (action.type) {
+    case ADD_RANGE_FILTER: {
+      const {filter} = action.payload;
+      const filters = insertItemToArray(shelfSpec.filters, 0, filter);
+      return {
+        ...shelfSpec,
+        filters: filters
+      };
+    }
+    case ADD_ONE_OF_FILTER: {
+      const {filter} = action.payload;
+      const filters = insertItemToArray(shelfSpec.filters, 0, filter);
+      return {
+        ...shelfSpec,
+        filters: filters
+      };
+    }
+    case REMOVE_RANGE_FILTER: {
```

Comment:
I think we can have just one `FILTER_REMOVE`  that takes `index` as parameter and remove the item from the array.  (And no need to do O(N) search.)